### PR TITLE
P1 - Asset listing [33] [Q]

### PR DESCRIPTION
**Pull request by Space-zee**

### CODE REVIEW

**Comment by Space-zee** [comment]
https://app.clickup.com/t/33yxp94

File: frontend/app/features/assets/controllers/NewAssetController.tsx [hunk 1]
```diff
@@ -1,46 +1,39 @@
...
+  const legalInfoForm = useAtomValue(newAssetModel.$legalInfoForm)
+  const investmentReturnForm = useAtomValue(newAssetModel.$investmentReturnForm)
+  const rentalManagementForm = useAtomValue(newAssetModel.$rentalManagementForm)
   const {address} = useAccount()
   const onFormChange = useSetAtom(newAssetModel.$onFormChange)
-  const onSubmit = useSetAtom(newAssetModel.$onSubmit)
-  const onMount = useSetAtom(newAssetModel.$onMount)
+  //const onMount = useSetAtom(newAssetModel.$onMount)
+  const [assetListingStatus, setAssetListingStatus] = useState<AssetListingStatus>(AssetListingStatus.welcome)
+  const [currentForm, setCurrentForm] = useState<UiForm<any>>(generalInfoForm)
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/controllers/NewAssetController.tsx:22]: please replace any with values type

File: frontend/app/features/assets/controllers/NewAssetController.tsx [hunk 2]
```diff
@@ -1,46 +1,39 @@
...
-            )
-            : "Please connect your wallet first"
-        }
-      </div>
-    </div>
+    address ? <AssetListing assetListingStatus={assetListingStatus} onChangeAssetListingStatus={setAssetListingStatus}
+                            onChange={onFormChange} generalInfoForm={generalInfoForm} legalInfoForm={legalInfoForm}
+                            investmentReturnForm={investmentReturnForm} rentalManagementForm={rentalManagementForm}
+                            returnHome={returnHome} currentForm={currentForm}
+                            onCurrentFormChange={setCurrentForm}/> : <><p>Please Connect your wallet!</p></>
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/controllers/NewAssetController.tsx:37]: Looks good but formatting is weird. What editor do you use?

File: frontend/app/features/assets/models/newAssetModel.ts [hunk 3]
```diff
@@ -1,88 +1,195 @@
...
-  name: '',
-  symbol: '',
-  title: '',
-  description: '',
-  status: AssetStatuses.upcoming,
-  tokenInfo_totalSupply: 10_000, // decimals = 0
-  tokenInfo_apr: 10, // percents
-  tokenInfo_tokenPriceDe6: 5, // decimals = 6
-  propertyInfo_images: 'https://ns.clubmed.com/dream/RESORTS_3T___4T/Asie_et_Ocean_indien/Bali/169573-1lng9n8nnf-swhr.jpg',
+const defaultAttrs = (): DefaultAssetListAttr => ({
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/models/newAssetModel.ts:14]: maybe 4 separated initializes is better?

File: frontend/app/features/assets/models/newAssetModel.ts [hunk 4]
```diff
@@ -1,88 +1,195 @@
...
+    bedrooms: 0,
+    roomsTotal: 0,
+    kitchens: 0,
+    livingRooms: 0,
+    terraces: 0,
+    balconies: 0,
+    garages: 0,
+    bathrooms: 0,
+    occupation: Occupation.occupied,
+    images: [''],
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/models/newAssetModel.ts:32]: I would advice to use an empty array as a default value. Or even null

File: frontend/app/features/assets/models/newAssetModel.ts [hunk 5]
```diff
@@ -1,88 +1,195 @@
...
+    llcFormationDocument: [''],
+    notaryConclusion: [''],
+    managementDocuments: [''],
+    rentalAgreement: [''],
+  },
+  investmentReturn: {
+    tokenName: '',
+    tokenSymbol: '',
+    tokenPrice: '',
+    tokenAmount: 0,
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/models/newAssetModel.ts:57]: Just a note: I am not sure this is a good UX to show 0 by default. Usually I do "null" in model and "" (empty string) on UI

File: frontend/app/features/assets/models/newAssetModel.ts [hunk 6]
```diff
@@ -1,88 +1,195 @@
...
+    listingFee: 0,
+    llcAdministrationFee: 0,
+    upfrontLlcFees: 0,
+  }
 })
 
 // stores
-export const $form = atom<UiNewAssetForm>({
-  values: defaultAttrs(),
+export const $generalInfoForm = atom<UiForm<any>>({
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/models/newAssetModel.ts:79]: any - can we replace with values type?

File: frontend/app/features/assets/views/AssetListing/index.tsx [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    if (assetListingStatus === AssetListingStatus.investmentAndReturn) {
+      onCurrentFormChange(investmentReturnForm)
+    }
+    if (assetListingStatus === AssetListingStatus.rentalAndManagement) {
+      onCurrentFormChange(rentalManagementForm)
+    }
+  }, [generalInfoForm.isValid, legalInfoForm.isValid, investmentReturnForm.isValid, rentalManagementForm.isValid])
+
+  const backButton = () => {
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/AssetListing/index.tsx:67]: backButton -> onBackButtonClick

File: frontend/app/shared/ui/views/BaseInput.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+type InputProps = {
+  title: string;
+  placeholder: string;
+  onChange: (formName: ListAssetsFormsNames, form: UiForm<any>, e: HTMLInputElement) => void;
+  form: UiForm<any>;
+  formType: ListAssetsFormsNames;
+  inputProps:  any;
+  classname?: string
+  inputType?: 'number' | 'string';
```
Comment (ruslanpalagin) [inline frontend/app/shared/ui/views/BaseInput.tsx:46]: I like this idea of inputProps proxying.
With this inputType is a part of inputProps, no?

File: frontend/app/shared/ui/views/BaseSelect.module.scss [hunk 9]
```diff
@@ -0,0 +1,62 @@
+.selectWrapper{
+  position: relative;
+}
+.title{
+  padding:0 0 8px 7px;
+  font-size: 12px;
+  color: var(--color-text-secondary);
+}
+.select{
```
Comment (ruslanpalagin) [inline frontend/app/shared/ui/views/BaseSelect.module.scss:9]: select - could it be removed from here?

File: frontend/app/shared/ui/views/BaseInput.tsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+import React, {useRef} from "react";
+import {ListAssetsFormsNames} from "../../../features/assets/types";
+import {UiForm} from "../../../../pkg/formType";
+//TODO: finish after MVP
+
+// type DropDownProps = {
+//   dropDownValuesArray: any[];
+// }
+//
+// const DropDown = (props: DropDownProps) => {
```
Comment (ruslanpalagin) [inline frontend/app/shared/ui/views/BaseInput.tsx:12]: Is this the time to remove comments?

File: frontend/app/shared/ui/views/BaseSelect.tsx [hunk 11]
```diff
@@ -0,0 +1,45 @@
+import s from './BaseSelect.module.scss'
+import {ArrowIconWithoutLine} from "../../../features/assets/views/ArrowIconWithoutLine";
```
Comment (ruslanpalagin) [inline frontend/app/shared/ui/views/BaseSelect.tsx:2]: Shared component MUST NOT import features.
Shared components MUST be reusable. 
So: 
1. please make all shared inputs feature-agnostic:
2. also (as a part of previous comments) - do not pass form and form type. You MAY pass just value/theme/onChange

File: frontend/app/shared/ui/views/TextArea.tsx [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+type Props = {
+  placeholder: string;
+  name: string;
+  title: string;
+  onChange: any;
+  inputProps: any;
+}
+export const TextArea = (props: Props) => {
+  const {placeholder, title, onChange, name, inputProps} = props
```
Comment (ruslanpalagin) [inline frontend/app/shared/ui/views/TextArea.tsx:16]: This is good. The rest shared components should be like this one

File: frontend/package.json [hunk 13]
```diff
@@ -21,6 +21,7 @@
     "@types/node": "18.11.9",
     "@types/react": "18.0.25",
     "@types/react-dom": "18.0.9",
+    "@types/react-select-country-list": "^2.2.1",
```
Comment (ruslanpalagin) [inline frontend/package.json:24]: Note: Adding a library we must analyze build size to make sure we do not exceed rational limits

File: frontend/pages/_app.tsx [hunk 14]
```diff
@@ -45,52 +48,53 @@ const MyApp: FC<Props> = ({Component, pageProps}) => {
...
-          {includeHotjarScript && (
-            <script dangerouslySetInnerHTML={{
-              __html: `
+              {/* <Navigation /> */}
+              {/* <Home data={''} /> */}
+              <Layout>
+                <Component {...pageProps} />
+              </Layout>
+              {showAdminMenu && <AdminMenu/>}
+            </ThemeProvider>
```
Comment (ruslanpalagin) [inline frontend/pages/_app.tsx:85]: Please adjust your editor to use .editorconfig (2 spaces, not 4)

**Review by ruslanpalagin — CHANGES_REQUESTED** [review]

File: frontend/app/features/assets/controllers/NewAssetController.tsx (hunk 1 above)
Comment (Space-zee) [reply frontend/app/features/assets/controllers/NewAssetController.tsx:22]: +

File: frontend/app/features/assets/views/AssetListing/index.tsx (hunk 7 above)
Comment (Space-zee) [reply frontend/app/features/assets/views/AssetListing/index.tsx:67]: +

File: frontend/app/shared/ui/views/BaseInput.tsx (hunk 8 above)
Comment (Space-zee) [reply frontend/app/shared/ui/views/BaseInput.tsx:46]: +

File: frontend/app/shared/ui/views/BaseSelect.module.scss (hunk 9 above)
Comment (Space-zee) [reply frontend/app/shared/ui/views/BaseSelect.module.scss:9]: +

File: frontend/app/shared/ui/views/BaseSelect.tsx (hunk 11 above)
Comment (Space-zee) [reply frontend/app/shared/ui/views/BaseSelect.tsx:2]: +

File: frontend/app/features/assets/views/AssetListing/RentalManagementForm/index.tsx [hunk 15]
```diff
@@ -0,0 +1,72 @@
+import s from "./styles.module.scss";
+import {UiForm} from "../../../../../../pkg/formType";
+import {BaseInput} from "../../../../../shared/ui/views/BaseInput";
+import {ListAssetsFormsNames, UiRentalManagementForm} from "../../../types";
+
+type Props = {
+  rentalManagementForm: UiRentalManagementForm;
+  inputProps: (form: UiForm<any>, name: string) => { name: string, value: any, "aria-invalid": string[] | undefined };
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/AssetListing/RentalManagementForm/index.tsx:8]: can we fix this any?
UiForm< any >

File: frontend/app/features/assets/views/AssetListing/index.tsx [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+import {RentalManagementForm} from "./RentalManagementForm";
+
+type Props = {
+  assetListingStatus: AssetListingStatus;
+  onChangeAssetListingStatus: (value: AssetListingStatus) => void;
+  generalInfoForm: UiGeneralInfoForm;
+  legalInfoForm: UiLegalInfoForm;
+  investmentReturnForm: UiInvestmentReturnForm;
+  rentalManagementForm: UiRentalManagementForm;
+  currentForm: UiForm<any>;
```
Comment (ruslanpalagin) [inline frontend/app/features/assets/views/AssetListing/index.tsx:30]: UiForm< any >

File: frontend/app/shared/ui/views/BaseSelect.tsx [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+export const BaseSelect = (props: Props) => {
+  const {placeholder, optionArray, inputProps, onChange, title, classname} = props;
+  const [closeDropdown, setCloseDropdown] = useState(true);
+  const inputRef = useRef<any>(null)
+  const showElem = (value: string | number) => {
+    // @ts-ignore
+    inputRef.current.value = `${value}`;
+    setCloseDropdown(true);
+    onChange(inputRef.current)
```
Comment (ruslanpalagin) [inline frontend/app/shared/ui/views/BaseSelect.tsx:25]: why do you need this ref at all?
could we simply emit new value?

File: frontend/app/shared/ui/views/TextArea.tsx [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+  return (
+    <div>
+      <p className={s.title}>{title}</p>
+      <textarea className={clsx(s.textArea, inputProps['aria-invalid'] && s.invalidInput)} placeholder={placeholder}
+                onChange={debounce(onChange, 1000)}
```
Comment (ruslanpalagin) [inline frontend/app/shared/ui/views/TextArea.tsx:21]: Note: I am afraid this is incorrect.
It's much safer to useCallback
this is a good article https://dmitripavlutin.com/react-throttle-debounce/

**Review by ruslanpalagin — CHANGES_REQUESTED** [review]

File: frontend/app/features/assets/views/AssetListing/index.tsx (hunk 16 above)
Comment (Space-zee) [reply frontend/app/features/assets/views/AssetListing/index.tsx:30]: +

**Review by ruslanpalagin — APPROVED** [review]

**Pull request merged**